File: server/app/cache.py
"""
Redis cache operations for ZenRay.
"""
import json
import logging
from typing import Any, Optional

import redis.asyncio as redis

from app.config import get_settings

logger = logging.getLogger("zenray.cache")

_redis: Optional[redis.Redis] = None

# Cache TTLs in seconds
CACHE_TTL_SHORT = 60       # 1 minute - for frequently changing data
CACHE_TTL_MEDIUM = 300     # 5 minutes - for run lists
CACHE_TTL_LONG = 3600      # 1 hour - for completed run details


# =============================================================================
# Connection Management
# =============================================================================

async def init_cache():
    """Initialize Redis connection."""
    global _redis
    settings = get_settings()
    _redis = redis.Redis(
        host=settings.redis_host,
        port=settings.redis_port,
        db=settings.redis_db,
        decode_responses=True,
    )
    await _redis.ping()
    logger.info("Redis connection initialized")


async def close_cache():
    """Close Redis connection."""
    global _redis
    if _redis:
        await _redis.close()
        _redis = None
        logger.info("Redis connection closed")


def get_redis() -> redis.Redis:
    """Get Redis client."""
    if _redis is None:
        raise RuntimeError("Redis not initialized")
    return _redis


# =============================================================================
# Cache Operations
# =============================================================================

async def cache_get(key: str) -> Optional[Any]:
    """Get value from cache."""
    r = get_redis()
    value = await r.get(key)
    if value:
        return json.loads(value)
    return None


async def cache_set(key: str, value: Any, ttl: int = CACHE_TTL_MEDIUM):
    """Set value in cache with TTL."""
    r = get_redis()
    await r.setex(key, ttl, json.dumps(value, default=str))


async def cache_delete(key: str):
    """Delete key from cache."""
    r = get_redis()
    await r.delete(key)


async def cache_delete_pattern(pattern: str):
    """Delete all keys matching pattern."""
    r = get_redis()
    cursor = 0
    while True:
        cursor, keys = await r.scan(cursor, match=pattern, count=100)
        if keys:
            await r.delete(*keys)
        if cursor == 0:
            break


# =============================================================================
# Cache Key Builders
# =============================================================================

def run_key(run_id: str) -> str:
    """Cache key for a run."""
    return f"run:{run_id}"


def run_detail_key(run_id: str) -> str:
    """Cache key for run detail."""
    return f"run_detail:{run_id}"


def step_key(step_id: str) -> str:
    """Cache key for a step."""
    return f"step:{step_id}"


def runs_list_key(
    pipeline_name: Optional[str] = None,
    status: Optional[str] = None,
    user_id: Optional[str] = None,
) -> str:
    """Cache key for runs list."""
    return f"runs_list:{user_id or 'all'}:{pipeline_name or 'all'}:{status or 'all'}"


# =============================================================================
# Invalidation Helpers
# =============================================================================

async def invalidate_run(run_id: str):
    """Invalidate all caches related to a run."""
    await cache_delete(run_key(run_id))
    await cache_delete(run_detail_key(run_id))
    await cache_delete_pattern("runs_list:*")


async def invalidate_step(step_id: str, run_id: str):
    """Invalidate step and related caches."""
    await cache_delete(step_key(step_id))
    await cache_delete(run_detail_key(run_id))

File: server/src/cache.ts
import Redis from 'ioredis'
import { config } from './config.js'

let redis: Redis | null = null

export const CACHE_TTL_SHORT = 60       // 1 minute
export const CACHE_TTL_MEDIUM = 300     // 5 minutes
export const CACHE_TTL_LONG = 3600      // 1 hour

export async function initCache(): Promise<void> {
  redis = new Redis({
    host: config.redis.host,
    port: config.redis.port,
    db: config.redis.db,
    maxRetriesPerRequest: 1,
    retryStrategy: () => null,
    lazyConnect: true,
  })
  redis.on('error', () => {}) // Suppress connection errors
  await redis.connect()
  await redis.ping()
}

export async function closeCache(): Promise<void> {
  if (redis) {
    await redis.quit()
    redis = null
  }
}

export function getRedis(): Redis {
  if (!redis) throw new Error('Redis not initialized')
  return redis
}

// Cache operations
export async function cacheGet<T>(key: string): Promise<T | null> {
  const r = getRedis()
  const value = await r.get(key)
  if (value) return JSON.parse(value) as T
  return null
}

export async function cacheSet(key: string, value: unknown, ttl: number = CACHE_TTL_MEDIUM): Promise<void> {
  const r = getRedis()
  await r.setex(key, ttl, JSON.stringify(value))
}

export async function cacheDelete(key: string): Promise<void> {
  const r = getRedis()
  await r.del(key)
}

export async function cacheDeletePattern(pattern: string): Promise<void> {
  const r = getRedis()
  let cursor = '0'
  do {
    const [newCursor, keys] = await r.scan(cursor, 'MATCH', pattern, 'COUNT', 100)
    cursor = newCursor
    if (keys.length > 0) {
      await r.del(...keys)
    }
  } while (cursor !== '0')
}

// Cache key builders
export function runKey(runId: string): string {
  return `run:${runId}`
}

export function runDetailKey(runId: string): string {
  return `run_detail:${runId}`
}

export function stepKey(stepId: string): string {
  return `step:${stepId}`
}

export function runsListKey(pipelineName?: string, status?: string): string {
  return `runs_list:${pipelineName || 'all'}:${status || 'all'}`
}

// Invalidation helpers
export async function invalidateRun(runId: string): Promise<void> {
  await cacheDelete(runKey(runId))
  await cacheDelete(runDetailKey(runId))
  await cacheDeletePattern('runs_list:*')
}

export async function invalidateStep(stepId: string, runId: string): Promise<void> {
  await cacheDelete(stepKey(stepId))
  await cacheDelete(runDetailKey(runId))
}

